### PR TITLE
Allow rulesets to specify custom song select filtering criteria

### PR DESCRIPTION
**Pull request by bdach**
Supersedes https://github.com/ppy/osu/pull/11925 + https://github.com/ppy/osu/pull/11926.

Intended towards https://github.com/ppy/osu/issues/5668.

This is mostly the branch [I linked in this comment](https://github.com/ppy/osu/pull/11925#issuecomment-789159734), with the last two commits being the sole difference. The goal of those commits is to expose a better interface of the parsing helpers to the end users (so that they don't have to call `TryParseFloat...` and then `TryUpdateCriteria...` as I had it initially in the "lazy exposition" version).

They're still a bit obtuse, unfortunately, as can be seen by the length of the xmldoc. I tried to explain to the best of my ability what those do, but I don't know if I've succeeded in that.

(On the upside, that change gets rid of some `case...when` usages, so maybe a plus?)

### CODE REVIEW

**Review by smoogipoo — COMMENTED** [review]
I really like this. Just one question.

File: osu.Game/Screens/Select/FilterControl.cs [hunk 1]
```diff
@@ -53,6 +59,8 @@ public FilterCriteria CreateCriteria()
             if (!maximumStars.IsDefault)
                 criteria.UserStarDifficulty.Max = maximumStars.Value;
 
+            criteria.RulesetCriteria = rulesets.GetRuleset(ruleset.Value.ID.Value).CreateInstance().CreateRulesetFilterCriteria();
```
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Screens/Select/FilterControl.cs:62]: That's a `RulesetInfo`, not a full ruleset.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Screens/Select/FilterControl.cs:62]: Oh wait never mind. Sure.

**Review by smoogipoo — APPROVED** [review]
This looks good to me. @peppy requesting your opinion too since you were vocal about the old revision.

**Review by peppy — APPROVED** [review]
Already checked over this one on the weekend. Looks fine 👍🏻

**Pull request merged**